File: rulebook-eslint/src/rules/lowercase-hexadecimal.ts
import { Rule } from 'eslint';
import { Literal } from 'estree';
import messages from '../messages.js';
import RulebookRule from './rulebook-rules.js';

/** {@link https://hanggrian.github.io/rulebook/rules/#lowercase-hexadecimal|See detail} */
class LowercaseHexadecimalRule extends RulebookRule {
    constructor() {
        super('lowercase-hexadecimal', {
            [LowercaseHexadecimalRule.MSG]: messages.get(LowercaseHexadecimalRule.MSG),
        });
    }

    override create(context: Rule.RuleContext) {
        return {
            Literal(node: Literal) {
                // checks for violation
                if (typeof node.value !== 'number') {
                    return;
                }
                const value = node.raw;
                if (value === undefined || !value.toLowerCase().startsWith('0x')) {
                    return;
                }
                const valueReplacement = value.toLowerCase();
                if (value === valueReplacement) {
                    return;
                }
                context.report({
                    node,
                    messageId: LowercaseHexadecimalRule.MSG,
                    data: {
                        $0: valueReplacement,
                    },
                });
            },
        };
    }

    static MSG: string = 'lowercase.hexadecimal';
}

export default new LowercaseHexadecimalRule();

File: rulebook-cppcheck/rulebook_cppcheck/checkers/lowercase_hexadecimal.py
from rulebook_cppcheck.checkers.rulebook_checkers import RulebookTokenChecker
from rulebook_cppcheck.messages import _Messages

try:
    from cppcheckdata import Token
except ImportError:
    from cppcheck.Cppcheck.addons.cppcheckdata import Token


class LowercaseHexadecimalChecker(RulebookTokenChecker):
    """See detail: https://hanggrian.github.io/rulebook/rules/#lowercase-hexadecimal"""
    ID: str = 'lowercase-hexadecimal'
    _MSG: str = 'lowercase.hexadecimal'

    def process_tokens(self, tokens: list[Token]) -> None:
        # checks for violation
        for token in [t for t in tokens if t.isNumber]:
            value: str = token.str
            if not value.lower().startswith('0x'):
                return
            value_replacement: str = value.lower()
            if value == value_replacement:
                return
            self.report_error(token, _Messages.get(self._MSG, value_replacement))
